core/jvm/src/LocalDate.kt

```diff
@@ -100,6 +102,8 @@ public actual class LocalDate internal constructor(internal val value: jtLocalDa
     @PublishedApi
     @JvmName("toEpochDays")
     internal fun toEpochDaysJvm(): Int = value.toEpochDay().clampToInt()
+
+    private fun writeReplace(): Any = Ser(Ser.DATE_TAG, this)
```

Comment:
The `java.time` classes also have a [`readObject` method](https://github.com/openjdk/jdk/blob/5dd0acb3cddb96845062c0b7cee1e384e69f43cb/src/java.base/share/classes/java/time/LocalDate.java#L2176-L2185) that always throws an exception with a comment that this is about defending against malicious streams. I'm not very familiar with Java's serialization, so I don't know if this is needed.

Reply:
I found [this post](https://nipafx.dev/java-serialization-proxy-pattern/) about serialization via `writeReplace`/`readResolve`. [This section](https://nipafx.dev/java-serialization-proxy-pattern/#artificial-byte-stream) is about the always throwing `readObject` method. Without it, a hand crafted byte stream could cause deserialization of e.g. `LocalDate` without going through `Ser.readResolve`. This could violate invariants of the classes. I think it could for example sneak in `null` to the non-nullable property `LocalDate.value`.

Reply:
Thanks! Yes, `deserialize("aced00057372001a6b6f746c696e782e6461746574696d652e4c6f63616c44617465618443f17dae33e70200014c00097472756556616c75657400154c6a6176612f74696d652f4c6f63616c446174653b787070"
                .hexToByteArray(HexFormat { bytes.byteSeparator = "" }))` does indeed create a `LocalDate` whose `value` is `null`.

Reply:
@lukellmann, would you like to make a PR that fixes (and tests!) this problem?

Reply:
sure, but it will probably take a few days

Reply:
I'm on it but it will take a bit (I'm going through the [spec](https://docs.oracle.com/en/java/javase/24/docs/specs/serialization/index.html)).

Reply:
I've started implementing it, still refining the tests.

Reply:
I've started implementing it, still refining the tests.

Reply:
#522

Sorry that it took so long, I was initially trying to do more than what's in that PR (testing that classes stick to a certain pattern when implementing `java.io.Serializable`, see the other branches in [my fork](https://github.com/lukellmann/kotlinx-datetime)). But I've now shrunk it down to the most important fixes and might open additional PRs later.